osu.Game/Beatmaps/BeatmapDifficultyCache.cs

```diff
@@ -302,29 +265,31 @@ protected override void Dispose(bool isDisposing)
 
         public readonly struct DifficultyCacheLookup : IEquatable<DifficultyCacheLookup>
         {
-            public readonly int BeatmapId;
-            public readonly int RulesetId;
-            public readonly Mod[] Mods;
+            public readonly BeatmapInfo Beatmap;
```

Comment:
I specifically didn't do this because I'm not sure we want to be storing a permanent game-wide reference to the `BeatmapInfo`. The same in a way with `RulesetInfo`, though I'm less concerned there.

It was already done this way (https://github.com/ppy/osu/commit/00e6217f60c9d1981e1eb16e1b21b39a70b844a9) before, and I don't remember if I actually noticed an issue before changing it to ints. Can this be changed back to storing the raw IDs, just to prevent unintentional regressions in an otherwise refactoring PR?

Reply:
I considered this, but I think it's fine. They are model types and shouldn't be an issue. Although I'm not sure how this will play in with realm.

Do you have an alternative? I added these in to provide a way to perform the construction/computation when required.

Reply:
As mentioned in discord, one way around this would be to add an extra parameter to GetAsync which pipes through the `TInstance` (or `object` if we don't want more generics) – is this a direction you'd be happy with?

Reply:
I think I'm just going to cautiously accept this as it is...

Reply:
Shouldn't be a huge deal if we find out this is too much of an overhead, at least.